### ISSUE TITLE
Signed Webhook Authentication Failing in .NET Core 5 Web API

### ISSUE THREAD

**Issue by enjoau** [sendgrid/sendgrid-csharp#1131]
Hi Group,

I have a .NET Core 5.0 Web API which listens for Signed WebHook events sent from SendGrid however I am unable to authenticate the signature against my WebHook Validation key.

I've read numerous posts on this subject on this forum which suggest I should place a line break & carriage return after the response body which I have done but this still doesn't work, authentication still fails.

This is the main thread I've been reading, #https://github.com/sendgrid/sendgrid-csharp/issues/1018 but I've also followed posts on the NodeJS and Java forums detailing the same issue and none of the solutions I've seen have fixed my issue, indeed, it looks as if very few posters have found truly satisfactory answers and the threads get closed before a definitive solution is outlined.

I'm currently using the native .NET Core ECDSA Cryptography classes for my verification but have also tried with the SendGrid NuGet Package for .Net Core  using [RequestValidator](https://github.com/sendgrid/sendgrid-csharp/blob/main/src/SendGrid/Helpers/EventWebhook/RequestValidator.cs) and the [StrongGrid](https://github.com/Jericho/StrongGrid) library, also from NuGet.

I'm posting my requests locally using both Postman and Fiddler and have tried with and without the "\r\n" tagged to the end of the request body.

My verification method is as follows:

```csharp
/// <summary>
/// Verifies the SendGrid digital signature
/// </summary>
/// <param name="request">The HTTP request</param>>
/// <returns>A boolean indicating whether this is a valid SendGrid request</returns>
public async Task<bool> VerifySigning(HttpRequest request)
{
     var publicKey = _configuration.GetValue<string>("AppSettings:SendGridPublicKey");
     var signature = request.Headers.FirstOrDefault(x => x.Key.ToLower() == "x-twilio-email-event-webhook-signature").Value.ToString();
     var timestamp = request.Headers.FirstOrDefault(x => x.Key.ToLower() == "x-twilio-email-event-webhook-timestamp").Value.ToString();

    string requestBody;
    using (var streamReader = new StreamReader(request.Body))
    {
        requestBody = await streamReader.ReadToEndAsync().ConfigureAwait(false);
    }

    requestBody += "\r\n"; //tried with and without this line

    var signatureBytes = Convert.FromBase64String(signature);
    var publicKeyBytes = Convert.FromBase64String(publicKey);

    var data = Encoding.UTF8.GetBytes($"{timestamp}{requestBody}");

    var eCDsa = ECDsa.Create();
    eCDsa.ImportSubjectPublicKeyInfo(publicKeyBytes, out _);
    var verified = eCDsa.VerifyData(data, signatureBytes, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);

    if (verified == false)
    {
        _logger.LogError("WebHook Signature Validation Failed.");
        _logger.LogInformation($"Host: {request.Host}");
        _logger.LogInformation($"User Agent: {request.Headers.FirstOrDefault(x => x.Key.ToLower() == "user-agent").Value}");

        await SendAuthenticationFailureEmail(request);

        return false;
    }

    return true;
}
```

My request is:

```
POST https://localhost:5001/api/sendgridwebhooks/Events/ HTTP/1.1
Content-Length: 1187
Accept: */*
Accept-Encoding: gzip, deflate, br
X-Twilio-Email-Event-Webhook-Timestamp: 1629430014
X-Twilio-Email-Event-Webhook-Signature: MEYCIQDmpczVVY+3iJOwEU3/Q4oH0HVNxWdHSR+fBffngR/iMwIhANh0dan2pDvdR7PTRWewAsC1J/3CxiptXyFF1NVsM3j7
Host: localhost:5001
User-Agent: Auth Test

[
    {
        "attempt": "49",
        "category": [
            "EDM"
        ],
        "email": "hello@georgina-rose.com",
        "event": "deferred",
        "ip": "149.72.187.208",
        "mc_stats": "singlesend",
        "phase_id": "send",
        "response": "Connection reset by peer",
        "send_at": "1629278400",
        "sg_event_id": "ZGVmZXJyZWQtNDktMTQwNTc2NDctM29zTzhlUjVTVktBNk1DTlRoX2xLUS0z",
        "sg_message_id": "3osO8eR5SVKA6MCNTh_lKQ.filterdrecv-7577c5795b-b96wd-1-611CD0D8-57.3",
        "sg_template_id": "d-673da88843d14122a7f3ddcda77738e4",
        "sg_template_name": "Clone: Version 2021-08-18T07:18:44.798Z 2021-08-18T08:18:55.192Z",
        "singlesend_id": "ee38f022-fffc-11eb-9e48-864d09ce8190",
        "singlesend_name": "180821_ENJO_T+S_EDM4_ENJOweekly",
        "smtp-id": "<3osO8eR5SVKA6MCNTh_lKQ@ismtpd0197p1mdw1.sendgrid.net>",
        "template_hash": "2f82ac6007c488b88dcbc6c5c1d6476a9213ab8a772640019259b33541467806",
        "template_id": "d-673da88843d14122a7f3ddcda77738e4",
        "template_version_id": "40c93b03-ad45-44b6-ad13-bc91cf6fd73b",
        "timestamp": 1629430008,
        "tls": 0
    }
]
```

Please could someone let me know what may be going wrong?

Many thanks,

Stuart

**Comment by shwetha-manvinkurke** [comment]
Can you provide a sample of your code where you tried with our `RequestValidator`?

**Comment by enjoau** [comment]
Hi @shwetha-manvinkurke,

Thanks for the reply.

Below is the code using your Request Validator. I have tried both with and without the "\r\n" appended to the Request Body.

```csharp
/// <summary>
/// Verifies the SendGrid digital signature
/// </summary>
/// <param name="request">The HTTP request</param>>
/// <returns>A boolean indicating whether this is a valid SendGrid request</returns>
public async Task<bool> VerifySigning(HttpRequest request)
{
	var publicKey = _configuration.GetValue<string>("AppSettings:SendGridPublicKey");
	var signature = request.Headers.FirstOrDefault(x => x.Key.ToLower() == "x-twilio-email-event-webhook-signature").Value.ToString();
	var timestamp = request.Headers.FirstOrDefault(x => x.Key.ToLower() == "x-twilio-email-event-webhook-timestamp").Value.ToString();

	string requestBody;
	using (var streamReader = new StreamReader(request.Body))
	{
		requestBody = await streamReader.ReadToEndAsync().ConfigureAwait(false);
	}

	requestBody += "\r\n";

	var requestValidator = new RequestValidator();
	var key = requestValidator.ConvertPublicKeyToECDSA(publicKey);

	var verified = requestValidator.VerifySignature(key, requestBody, signature, timestamp);

	if (verified == false)
	{
		_logger.LogError("WebHook Signature Validation Failed.");
		_logger.LogInformation($"Host: {request.Host}");
		_logger.LogInformation($"User Agent: {request.Headers.FirstOrDefault(x => x.Key.ToLower() == "user-agent").Value}");

		await SendAuthenticationFailureEmail(request);

		return false;
	}

	return true;
}
```

Hope this helps.

Cheers

**Comment by shwetha-manvinkurke** [comment]
Hmm sendgrid expects a a `\r\n` after each event in case of a multi event payload in addition to the trailing one, but that doesn't apply to your particular use case I am guessing from your example. Have you tried logging your `requestBody` to see if that is as expected with no trailing and leading whitespaces, etc.?

**Comment by enjoau** [comment]
Hi @shwetha-manvinkurke,

As per your suggestion I have logged the Request Body and the log is attached.

I can't see any leading or trailing whitespaces but to be doubly sure I added a Trim() function to the request body after logging and before adding the "\r\n" which has made no difference. The authentication still fails.

```csharp
string requestBody;
using (var streamReader = new StreamReader(request.Body))
{
	requestBody = await streamReader.ReadToEndAsync().ConfigureAwait(false);
}

_logger.LogInformation(requestBody);

requestBody = requestBody.Trim();
requestBody += "\r\n";
```

[ex20210922.log](https://github.com/sendgrid/sendgrid-csharp/files/7206904/ex20210922.log)
s.

Cheers

**Comment by enjoau** [comment]
Hi @shwetha-manvinkurke,

I have solved the problem.

After following your suggestion of logging the Request Body it became apparent that the JSON had been beautified by Postman and contained a bunch of formatting such as additional "\r\n" as well as spaces and tabs which was being posted instead of the raw JSON as one continuous line of data.

This was obviously completely throwing the Elliptic Curve verifier.
Once I posted the raw request to the service it verified fine.

School boy error and therein lies a lesson. Do not allow JSON to be beautified before posting.
Who would have thought something so innocuous could be such a problem.

Thanks for your help
